Code_Exercises/Functors/solution.cpp

```diff
@@ -0,0 +1,235 @@
+    exit(1);
+  }
+
+  int size = width * 4;
+
+  float *filterData = new float[size];
+
+  for (int i = 0; i < width; ++i)
+  {
+    auto index = i * 4;
+    auto isCenter = (i == (width / 2));
+
+    filterData[index + 0] = 1.0f / static_cast<float>(width);
+    filterData[index + 1] = 1.0f / static_cast<float>(width);
+    filterData[index + 2] = 1.0f / static_cast<float>(width);
+    filterData[index + 3] = isCenter ? 1.0f : 0.0f;
+  }
+
+  return util::image_ref<float>{filterData, width, 1, 4, 0};
+}
+
+inline constexpr util::filter_type filterType = util::filter_type::blur;
+inline constexpr int filterWidth = 11;
+inline constexpr int halo = filterWidth / 2;
+
+TEST_CASE("image_convolution_1D", "1D_solution")
+{
+  const char *inputImageFile =
+      "../Images/dogs.png";
+  const char *outputImageFile =
+      "../Images/blurred_dogs_1D.png";
+
+  auto inputImage = util::read_image(inputImageFile, halo);
+
+  auto outputImage = util::allocate_image(
+      inputImage.width(), inputImage.height(), inputImage.channels());
+
+  auto filter = linear_blur(filterWidth);
+
+  try
+  {
+    sycl::queue myQueue{sycl::cpu_selector_v};
+
+    std::cout << "Running on "
+              << myQueue.get_device().get_info<sycl::info::device::name>()
+              << "\n";
+
+    auto inputImgWidth = inputImage.width();
+    auto inputImgHeight = inputImage.height();
+    auto channels = inputImage.channels();
+    auto filterWidth = filter.width();
+    auto halo = filter.half_width();
+
+    auto globalRange = sycl::range(inputImgWidth, inputImgHeight);
+    auto localRange = sycl::range(1, 32);
+    auto ndRange = sycl::nd_range(globalRange, localRange);
+
+    auto inBufRange =
+        sycl::range(inputImgHeight + (halo * 2), inputImgWidth + (halo * 2)) *
+        sycl::range(1, channels);
+    auto outBufRange =
+        sycl::range(inputImgHeight, inputImgWidth) * sycl::range(1, channels);
+    auto filterRange = sycl::range(filterWidth * channels);
+
+    {
+      auto inBuf = sycl::buffer{inputImage.data(), inBufRange};
+      auto outBuf = sycl::buffer<float, 2>{outBufRange};
+      auto tempBuf = sycl::buffer<float, 2>{inBufRange};
+      auto filterBuf = sycl::buffer{filter.data(), filterRange};
+      outBuf.set_final_data(outputImage.data());
+
+      auto inBufVec = inBuf.reinterpret<sycl::float4>(inBufRange /
+                                                      sycl::range(1, channels));
+      auto tempBufVec = tempBuf.reinterpret<sycl::float4>(
+          inBufRange / sycl::range(1, channels));
+      auto outBufVec = outBuf.reinterpret<sycl::float4>(
+          outBufRange / sycl::range(1, channels));
+      auto filterBufVec = filterBuf.reinterpret<sycl::float4>(
+          filterRange / channels);
+
+
+
+      util::benchmark(
+          [&]()
+          {
+            myQueue.submit([&](sycl::handler &cgh) {
+              ImageConvolutionFunctor<sycl::float4>convolve(cgh, inBufVec, tempBufVec,
```

Comment:
That doesn't look like it will compile?

Reply:
It does compile. Why do you think it won't?

Reply:
Usually you need a “.” to access member functions

Reply:
This is correct, but it would be good to get a bit of clang-formatting on it to make this a bit clearer.